File: 0145_rev_nums/program.py
# PROJECT EULER 145 MAIN PROGRAM
# - James deLeon

get_rev: int = lambda n: int("".join([i for i in str(n)][::-1]))

def is_reversible(n: int) -> bool:
    rev_num: int = n + get_rev(n)
    rev_arr: list = [int(i) for i in str(rev_num)]
    all_odd: bool = True
    for digit in rev_arr:
        if digit % 2 == 0:
            return False
    return True

if __name__ == "__main__":
    acc: int = 0
    for i in range(1, 1000000000):
        #- add extra condition for numbers that end in zero, since no leading 
        # zeroes are allowed for its reversed counterpart
        if is_reversible(i) and (i % 10 != 0):
            acc += 1
        if i % 50000 == 0: print("{0}: {1}".format(i, acc))
    print("reversible count: {}".format(acc))

File: 0145_rev_nums/program.c
// PROJECT EULER 145 MAIN PROGRAM
// - James deLeon (translated to C)

#include <stdio.h>
#include <stdlib.h>

#define LIMIT 1000000000

int reverse(int n);
int is_reversible(int n);

int main(int argc, char* argv[])
{
    int acc = 0;
    for (int i = 1; i < LIMIT; i++) {
        if(i % 10 != 0 && is_reversible(i))
        {
            acc++;
        }
        if(i % 50000 == 0)
        {
            printf("%d: %d\n", i, acc);
        }
    }
    printf("reversible count: %d\n", acc);
    return 0;
}

int reverse(int n)
{
    int rev = 0;
    while(n > 0)
    {
        rev = rev * 10 + n % 10;
        n /= 10;
    }
    return rev;
}

int is_reversible(int n)
{
    int sum = n + reverse(n);
    while (sum > 0)
    {
        int digit = sum % 10;
        if(digit % 2 == 0)
        {
            return 0;  // Even digit found
        }
        sum /= 10;
    }
    return 1;  // All digits are odd
}
